File: supabase/functions/chatbot-whatsapp-webhook/chatbot/main_menu.ts
import { ConversationState, updateConversationState } from './conversation_state.ts'
import { getOrCreateCustomer, getCartItems, getCartSummary, getCustomerAddresses, getActiveOrdersCount, getCustomerProfile, getCustomerPaymentMethods } from './database_helpers.ts'
import { formatMainMenu, formatOrderTypeSelection, formatNewOrderWarning, formatCartView, formatError, formatAddressManagement, formatProfileView, formatBillingManagement } from './message_formatter.ts'
import { handleReviewFlow } from './review_system.ts'

export async function handleMainMenu(
  chatId: string,
  customerId: string,
  messageBody: string,
  state: ConversationState
): Promise<string> {
  const input = messageBody.trim().toLowerCase()
  
  // Check for existing cart across all locations
  let existingCartLocation = null
  let cartItems: any[] = []
  
  // Find any existing cart (check all possible locations or use a general query)
  const allCartItems = await getAllCartItemsForCustomer(customerId)
  if (allCartItems.length > 0) {
    existingCartLocation = allCartItems[0].location_id
    cartItems = allCartItems
  }
  
  const cartCount = cartItems.length
  
  // Get active orders count
  const activeOrderCount = await getActiveOrdersCount(customerId)
  
  // Get location name for cart display
  let cartLocationName = ''
  if (existingCartLocation) {
    const locationInfo = await getLocationInfo(existingCartLocation)
    cartLocationName = locationInfo?.name || ''
  }
  
  // Handle different menu options
  switch (input) {
    case '1':
      // Continue with existing cart (only if cart exists)
      if (cartCount > 0) {
        const cartSummary = await getCartSummary(customerId, existingCartLocation!)
        await updateConversationState(chatId, {
          ...state,
          step: 'cart_view',
          selected_location_id: existingCartLocation,
          previous_step: 'main_menu'
        })
        return formatCartView(cartItems, cartSummary)
      } else {
        return formatMainMenu(undefined, cartCount, activeOrderCount, cartLocationName)
      }
      
    case 'a':
    case 'make order':
    case 'order':
    case 'make new order':
      // Check if there's an existing cart
      if (cartCount > 0) {
        await updateConversationState(chatId, {
          ...state,
          step: 'new_order_warning',
          previous_step: 'main_menu'
        })
        return formatNewOrderWarning(cartCount, cartLocationName)
      } else {
        await updateConversationState(chatId, {
          ...state,
          step: 'order_type',
          previous_step: 'main_menu'
        })
        return formatOrderTypeSelection()
      }
      
    case 'b':
    case 'previous orders':
    case 'orders':
      // Handle previous orders view
      return await handlePreviousOrders(chatId, customerId, state)
      
    case 'c':
    case 'profile':
    case 'my profile':
      // Handle profile view
      return await handleProfile(chatId, customerId, state)
      
    case 'd':
    case 'billing':
      // Handle billing view
      return await handleBilling(chatId, customerId, state)
      
    case 'e':
    case 'addresses':
      // Handle addresses view
      return await handleAddresses(chatId, customerId, state)
      
    default:
      // Check if it's a number for review flow compatibility
      const selectedNumber = parseInt(input)
      if (!isNaN(selectedNumber)) {
        // Try to handle as review flow
        return await handleReviewFlow(chatId, customerId, messageBody, state)
      }
      
      // Default: show main menu
      return formatMainMenu(undefined, cartCount, activeOrderCount, cartLocationName)
  }
}

export async function handleNewOrderWarning(
  chatId: string,
  customerId: string,
  messageBody: string,
  state: ConversationState
): Promise<string> {
  const selectedNumber = parseInt(messageBody.trim())
  
  // Get existing cart info
  const allCartItems = await getAllCartItemsForCustomer(customerId)
  const cartCount = allCartItems.length
  const existingCartLocation = allCartItems.length > 0 ? allCartItems[0].location_id : null
  
  // Get location name for display
  let cartLocationName = ''
  if (existingCartLocation) {
    const locationInfo = await getLocationInfo(existingCartLocation)
    cartLocationName = locationInfo?.name || ''
  }
  
  if (!isNaN(selectedNumber)) {
    switch (selectedNumber) {
      case 1: // Continue with current plate
        if (existingCartLocation) {
          const cartSummary = await getCartSummary(customerId, existingCartLocation)
          await updateConversationState(chatId, {
            ...state,
            step: 'cart_view',
            selected_location_id: existingCartLocation,
            previous_step: 'new_order_warning'
          })
          return formatCartView(allCartItems, cartSummary)
        } else {
          return formatMainMenu(undefined, 0, 0)
        }
        
      case 2: // Delete plate & make new order
        if (existingCartLocation) {
          // Clear the existing cart
          await clearCustomerCart(customerId, existingCartLocation)
        }
        await updateConversationState(chatId, {
          ...state,
          step: 'order_type',
          previous_step: 'new_order_warning'
        })
        return formatOrderTypeSelection()
        
      case 3: // View current plate
        if (existingCartLocation) {
          const cartSummary = await getCartSummary(customerId, existingCartLocation)
          // Don't change step, just show cart then return to warning
          return formatCartView(allCartItems, cartSummary) + '\n\n' + formatNewOrderWarning(cartCount, cartLocationName)
        } else {
          return formatMainMenu(undefined, 0, 0)
        }
        
      case 4: // Back to main menu
        await updateConversationState(chatId, {
          ...state,
          step: 'main_menu',
          previous_step: 'new_order_warning'
        })
        return formatMainMenu(undefined, cartCount, 0, cartLocationName)
        
      default:
        return formatError('Please select option 1, 2, 3, or 4')
    }
  }
  
  return formatError('Please select a valid option (1-4)')
}

async function getAllCartItemsForCustomer(customerId: string) {
  // This is a simplified version - in the actual implementation you'd query all cart items
  // regardless of location to find any existing cart
  const { data: cartItems, error } = await import('./database_helpers.ts')
    .then(module => module.getCartItems(customerId, '')) // Empty location to get all
    .catch(() => [])
  
  return cartItems || []
}

async function getLocationInfo(locationId: string) {
  try {
    const { createClient } = await import("https://esm.sh/@supabase/supabase-js@2")
    const supabase = createClient(
      Deno.env.get('SUPABASE_URL') ?? '',
      Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') ?? ''
    )
    
    const { data: location, error } = await supabase
      .from('locations')
      .select('id, name')
      .eq('id', locationId)
      .single()
      
    if (error) {
      console.error('Error getting location info:', error)
      return null
    }
    
    return location
  } catch (error) {
    console.error('Error in getLocationInfo:', error)
    return null
  }
}

async function clearCustomerCart(customerId: string, locationId: string) {
  try {
    const { createClient } = await import("https://esm.sh/@supabase/supabase-js@2")
    const supabase = createClient(
      Deno.env.get('SUPABASE_URL') ?? '',
      Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') ?? ''
    )
    
    const { error } = await supabase
      .from('cart_items')
      .delete()
      .eq('customer_id', customerId)
      .eq('location_id', locationId)
      
    if (error) {
      console.error('Error clearing cart:', error)
      return false
    }
    
    return true
  } catch (error) {
    console.error('Error in clearCustomerCart:', error)
    return false
  }
}

async function handlePreviousOrders(chatId: string, customerId: string, state: ConversationState): Promise<string> {
  // TODO: Implement previous orders view
  // For now, redirect to review flow for compatibility
  return await handleReviewFlow(chatId, customerId, 'previous orders', state)
}

async function handleProfile(chatId: string, customerId: string, state: ConversationState): Promise<string> {
  const profile = await getCustomerProfile(customerId)
  if (!profile) {
    return formatError('Unable to load profile. Please try again.')
  }
  
  await updateConversationState(chatId, {
    ...state,
    step: 'profile_view',
    previous_step: 'main_menu'
  })
  return formatProfileView(profile)
}

async function handleBilling(chatId: string, customerId: string, state: ConversationState): Promise<string> {
  const paymentMethods = await getCustomerPaymentMethods(customerId)
  
  await updateConversationState(chatId, {
    ...state,
    step: 'billing_list',
    previous_step: 'main_menu'
  })
  return formatBillingManagement(paymentMethods)
}

async function handleAddresses(chatId: string, customerId: string, state: ConversationState): Promise<string> {
  const addresses = await getCustomerAddresses(customerId)
  
  await updateConversationState(chatId, {
    ...state,
    step: 'address_list',
    previous_step: 'main_menu'
  })
  
  return formatAddressManagement(addresses)
} 

File: backend/internal/chatbot/main_menu.go
package chatbot

import (
	"context"
	"log"
	"strings"
)

func (s *Service) handleMainMenu(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	input := strings.ToLower(strings.TrimSpace(messageBody))

	allCartItems := s.getAllCartItemsForCustomer(ctx, customerID)
	existingCartLocation := ""
	var cartItems []CartItem
	if len(allCartItems) > 0 {
		existingCartLocation = allCartItems[0].LocationID
		cartItems = allCartItems
	}
	cartCount := len(cartItems)

	activeOrderCount := s.getActiveOrdersCount(ctx, customerID)

	cartLocationName := ""
	if existingCartLocation != "" {
		locationInfo := s.getLocationInfo(ctx, existingCartLocation)
		if locationInfo != nil {
			cartLocationName = locationInfo.Name
		}
	}

	switch input {
	case "1":
		if cartCount > 0 {
			cartSummary := s.getCartSummary(ctx, customerID, existingCartLocation)
			newState := state
			newState.Step = "cart_view"
			newState.SelectedLocationID = existingCartLocation
			newState.PreviousStep = "main_menu"
			s.updateConversationState(ctx, chatID, newState)
			return formatCartView(cartItems, cartSummary)
		}
		return formatMainMenu("", cartCount, activeOrderCount, cartLocationName)

	case "a", "make order", "order", "make new order":
		if cartCount > 0 {
			newState := state
			newState.Step = "new_order_warning"
			newState.PreviousStep = "main_menu"
			s.updateConversationState(ctx, chatID, newState)
			return formatNewOrderWarning(cartCount, cartLocationName)
		}
		newState := state
		newState.Step = "order_type"
		newState.PreviousStep = "main_menu"
		s.updateConversationState(ctx, chatID, newState)
		return formatOrderTypeSelection()

	case "b", "previous orders", "orders":
		return s.handlePreviousOrders(ctx, chatID, customerID, state, messageBody)

	case "c", "profile", "my profile":
		return s.handleProfileMenu(ctx, chatID, customerID, state)

	case "d", "billing":
		return s.handleBillingMenu(ctx, chatID, customerID, state)

	case "e", "addresses":
		return s.handleAddressesMenu(ctx, chatID, customerID, state)

	default:
		if _, hasNum := parseIntTrim(input); hasNum {
			return s.handleReviewFlow(ctx, chatID, customerID, messageBody, state)
		}
		return formatMainMenu("", cartCount, activeOrderCount, cartLocationName)
	}
}

func (s *Service) handleNewOrderWarning(ctx context.Context, chatID, customerID, messageBody string, state ConversationState) string {
	selectedNumber, hasNum := parseIntTrim(messageBody)
	allCartItems := s.getAllCartItemsForCustomer(ctx, customerID)
	cartCount := len(allCartItems)
	existingCartLocation := ""
	if cartCount > 0 {
		existingCartLocation = allCartItems[0].LocationID
	}

	cartLocationName := ""
	if existingCartLocation != "" {
		locationInfo := s.getLocationInfo(ctx, existingCartLocation)
		if locationInfo != nil {
			cartLocationName = locationInfo.Name
		}
	}

	if !hasNum {
		return formatError("Please select a valid option (1-4)")
	}

	switch selectedNumber {
	case 1:
		if existingCartLocation != "" {
			cartSummary := s.getCartSummary(ctx, customerID, existingCartLocation)
			newState := state
			newState.Step = "cart_view"
			newState.SelectedLocationID = existingCartLocation
			newState.PreviousStep = "new_order_warning"
			s.updateConversationState(ctx, chatID, newState)
			return formatCartView(allCartItems, cartSummary)
		}
		return formatMainMenu("", 0, 0, "")
	case 2:
		if existingCartLocation != "" {
			s.clearCustomerCart(ctx, customerID, existingCartLocation)
		}
		newState := state
		newState.Step = "order_type"
		newState.PreviousStep = "new_order_warning"
		s.updateConversationState(ctx, chatID, newState)
		return formatOrderTypeSelection()
	case 3:
		if existingCartLocation != "" {
			cartSummary := s.getCartSummary(ctx, customerID, existingCartLocation)
			return formatCartView(allCartItems, cartSummary) + "\n\n" + formatNewOrderWarning(cartCount, cartLocationName)
		}
		return formatMainMenu("", 0, 0, "")
	case 4:
		newState := state
		newState.Step = "main_menu"
		newState.PreviousStep = "new_order_warning"
		s.updateConversationState(ctx, chatID, newState)
		return formatMainMenu("", cartCount, 0, cartLocationName)
	default:
		return formatError("Please select option 1, 2, 3, or 4")
	}
}

func (s *Service) getAllCartItemsForCustomer(ctx context.Context, customerID string) []CartItem {
	return s.getCartItems(ctx, customerID, "")
}

func (s *Service) getLocationInfo(ctx context.Context, locationID string) *Location {
	l := &Location{}
	err := s.pool.QueryRow(ctx,
		`SELECT id, name FROM locations WHERE id = $1`,
		locationID,
	).Scan(&l.ID, &l.Name)
	if err != nil {
		log.Printf("Error getting location info: %v", err)
		return nil
	}
	return l
}

func (s *Service) clearCustomerCart(ctx context.Context, customerID, locationID string) bool {
	_, err := s.pool.Exec(ctx,
		`DELETE FROM cart_items WHERE customer_id = $1 AND location_id = $2`,
		customerID, locationID,
	)
	if err != nil {
		log.Printf("Error clearing cart: %v", err)
		return false
	}
	return true
}

func (s *Service) handlePreviousOrders(ctx context.Context, chatID, customerID string, state ConversationState, messageBody string) string {
	// TS redirects to review flow for compatibility.
	return s.handleReviewFlow(ctx, chatID, customerID, "previous orders", state)
}

func (s *Service) handleProfileMenu(ctx context.Context, chatID, customerID string, state ConversationState) string {
	profile := s.getCustomerProfile(ctx, customerID)
	if profile == nil {
		return formatError("Unable to load profile. Please try again.")
	}
	newState := state
	newState.Step = "profile_view"
	newState.PreviousStep = "main_menu"
	s.updateConversationState(ctx, chatID, newState)
	return formatProfileView(profile)
}

func (s *Service) handleBillingMenu(ctx context.Context, chatID, customerID string, state ConversationState) string {
	paymentMethods := s.getCustomerPaymentMethods(ctx, customerID)
	newState := state
	newState.Step = "billing_list"
	newState.PreviousStep = "main_menu"
	s.updateConversationState(ctx, chatID, newState)
	return formatBillingManagement(paymentMethods)
}

func (s *Service) handleAddressesMenu(ctx context.Context, chatID, customerID string, state ConversationState) string {
	addresses := s.getCustomerAddresses(ctx, customerID)
	newState := state
	newState.Step = "address_list"
	newState.PreviousStep = "main_menu"
	s.updateConversationState(ctx, chatID, newState)
	return formatAddressManagement(addresses)
}
